das/algorithms/3DES.py

```diff
@@ -0,0 +1,324 @@
+
+
+def dec2bin(num):
+
+    res = bin(num).replace("0b", "")
+
+    if len(res) % 4 != 0:
+        div = len(res) / 4
+        div = int(div)
+        counter = (4 * (div + 1)) - len(res)
+        for i in range(0, counter):
+            res = '0' + res
+    return res
+
+
+def bin2dec(binary):
+    return int(binary, 2)
+
+
+def hex2bin(s):
+    map_of_h2b = {'0': "0000",
+                  '1': "0001",
+                  '2': "0010",
+                  '3': "0011",
+                  '4': "0100",
+                  '5': "0101",
+                  '6': "0110",
+                  '7': "0111",
+                  '8': "1000",
+                  '9': "1001",
+                  'A': "1010",
+                  'B': "1011",
+                  'C': "1100",
+                  'D': "1101",
+                  'E': "1110",
+                  'F': "1111"}
+
+    res = ""
+    for i in range(len(s)):
+        res = res + map_of_h2b[s[i]]
+
+    return res
+
+
+def bin2hex(s):
+    h = {"0000": '0',
+                  "0001": '1',
+                  "0010": '2',
+                  "0011": '3',
+                  "0100": '4',
+                  "0101": '5',
+                  "0110": '6',
+                  "0111": '7',
+                  "1000": '8',
+                  "1001": '9',
+                  "1010": 'A',
+                  "1011": 'B',
+                  "1100": 'C',
+                  "1101": 'D',
+                  "1110": 'E',
+                  "1111": 'F'}
+
+    res = "".join(h[s[i:i+4]] for i in range(0, len(s), 4))
+    return res
+
+
+def LS(bit_string, num_of_shifts):
+    s = ""
+    for i in range(num_of_shifts):
+        for j in range(1, len(bit_string)):
+            s = s + bit_string[j]
+        s = s + bit_string[0]
+        bit_string = s
+        s = ""
+    return bit_string
+
+
+def RS(bit_string, num_of_shifts):
+    for i in range(num_of_shifts):
+        bit_string = bit_string[-1] + bit_string[:-1]
+
+    return bit_string
+
+
+def xor(a, b):
+    result = ""
+    for i in range(len(a)):
+        if a[i] == b[i]:
+            result = result + "0"
+        else:
+            result = result + "1"
+    return result
+
+
+def get_round_key_encrypt(key):
+    round_keys = []
+
+    # PC-1
+    key = "".join(key[i-1] for i in keyp)
+
+    # C0
+    left_key = key[:28]
+    # D0
+    right_key = key[28:]
+    for round_count in range(16):
+        left_key = LS(left_key, shift_table[round_count])
+        right_key = LS(right_key, shift_table[round_count])
+
+        k = left_key + right_key
+        round_keys.append("".join(k[i-1] for i in key_comp))
+
+    return round_keys
+
+
+def get_round_keys_decrypt(key):
+    round_keys = []
+
+    # PC-1
+    key = "".join(key[i-1] for i in keyp)
+
+    # C0
+    left_key = key[:28]
+    # D0
+    right_key = key[28:]
+    for round_count in range(15,-1,-1):
+        k = left_key + right_key
+        round_keys.append("".join(k[i-1] for i in key_comp))
```

Comment:
I replaced all instances of `permute` with this string comprehension line because I wasn't convinced we needed the `length` variable. As it turns out, we did not need it, but it'd probably be good to make a new `permute` function which does the `return "".join(s[i-1] for i in table)` for readability.